Yago/step11 ADD: Store the characters in a database using Room

YagoAccenture opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Integrated Room database for efficient local data management.
	- Added `Character` and `Episode` entities with persistent storage capabilities.
	- Introduced DAO interfaces for managing database interactions.
	- Implemented remote and local data source interfaces to streamline data retrieval processes.
	- Introduced a new `CharacterEntity` data model for improved data handling.
	- Added `EpisodeRemoteEntity` and `CharacterRemoteEntity` for structured remote data management.
	- Added custom exceptions for improved error reporting during data fetching.

- **Improvements**
	- Enhanced error handling and data fetching logic, including fallback mechanisms for local data retrieval.
	- Improved data handling in the application by optimizing the interaction between API and local database sources.
	- Updated character detail presentation to simplify data access.
	- Enhanced the architecture to support better separation of concerns and modular data management.
	- Streamlined data transformation processes between remote and local data models.

- **Chores**
	- Added necessary Room dependencies and Kotlin annotation processing support in the project configuration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
Step 11 ADD:
store the characters in the database using Room

Commit:
Step 11

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/AppDatabase.kt:
```diff
@@ -0,0 +1,31 @@
+import android.content.Context
+
+@Database(entities = [Character::class, Episode::class], version = 1)
+@TypeConverters(UbicationConverter::class, StringListConverter::class)
```
avoid to use typeconverters

YagoAccenture replied:
Fixed in next commit.

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/Character.kt:
```diff
@@ -16,11 +21,12 @@ data class Ubication(
     val url: String
 ) : Serializable
 
+@Entity(tableName = "episodes")
```
this table is not needed, just show an error if the episode details cannot be loaded.

YagoAccenture replied:
Fixed in next commit

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/CharacterDao.kt:
```diff
@@ -0,0 +1,18 @@
+    @Query("SELECT * FROM characters")
+    suspend fun getAllCharacters(): List<Character>
+
+    @Query("SELECT * FROM characters WHERE name = :name")
```
not needed

YagoAccenture replied:
Fixed in next commit

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/DetailsActivity.kt:
```diff
@@ -10,6 +10,8 @@ import com.squareup.picasso.Picasso
 
 class DetailsActivity : AppCompatActivity(), DetailsViewTranslator {
 
+    private lateinit var viewModel: DetailsViewModel
```
forbidden

YagoAccenture replied:
Fixed in next commit

YagoAccenture commented on app/src/main/java/com/mstudio/superheroarch/DetailsActivity.kt:
```diff
@@ -10,6 +10,8 @@ import com.squareup.picasso.Picasso
 
 class DetailsActivity : AppCompatActivity(), DetailsViewTranslator {
 
+    private lateinit var viewModel: DetailsViewModel
+
```
Fixed in next comit

YagoAccenture commented on app/src/main/java/com/mstudio/superheroarch/RickAndMortyRepository.kt:
```diff
@@ -1,43 +1,52 @@
-                    }
+        return try {
+            val response = apiRick.getEpisode(episodeUrl)
+            if (response.isSuccessful) {
+                val episode = response.body()
+                if (episode != null) {
+                    episodeDao.insertEpisode(episode)
+                    Result.success(episode)
                 } else {
-                    Result.failure(Exception("Error fetching episode details"))
+                    Result.failure(Exception("Episode not found"))
                 }
-            } catch (e: Exception) {
+            } else {
+                Result.failure(Exception("Error fetching episode details"))
+            }
+        } catch (e: Exception) {
+            val episodeCode = episodeUrl.split("/").last()
+            val localEpisode = episodeDao.getEpisodeByCode(episodeCode)
+            if (localEpisode != null) {
+                Result.success(localEpisode)
+            } else {
```
Fixed in next commit.

Commit:
Step 11 EDIT Changes due to comments:
ViewModel variable is no longer lateinit.
Removed the episode table from the entire app.

Commit:
Step 11 EDIT:
Removed the use of TypeConverters

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/Character.kt:
```diff
@@ -1,15 +1,34 @@
+    val locationName: String,
+    val firstEpisode: String
+) : Serializable {
+    constructor(
```
why do you need this?

YagoAccenture replied:
It has been replaced by a mapping

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/Character.kt:
```diff
@@ -1,15 +1,34 @@
+@Entity(tableName = "characters")
 data class Character(
-    val name: String,
+    @PrimaryKey val name: String,
```
a name cannot be a primarykey, you don't know if there are several characters with same name.

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/Character.kt:
```diff
@@ -1,15 +1,34 @@
+import androidx.room.PrimaryKey
 import java.io.Serializable
 
+@Entity(tableName = "characters")
```
you don't need to put the tableName, just @Entity is enough

YagoAccenture replied:
Fixed in next comit

PabloGrelaM commented on app/build.gradle.kts:
```diff
@@ -42,8 +42,12 @@ dependencies {
+    implementation(libs.room.runtime)
+    implementation(libs.room.ktx)
     implementation(libs.androidx.recyclerview)
+    implementation ("androidx.lifecycle:lifecycle-viewmodel-ktx:2.8.4")
```
keep consistency, this is not the way to declare dependencies

YagoAccenture replied:
Fixed in next comit

PabloGrelaM commented on app/build.gradle.kts:
```diff
@@ -1,6 +1,7 @@
 plugins {
     alias(libs.plugins.android.application)
     alias(libs.plugins.jetbrains.kotlin.android)
+    id("kotlin-kapt")
```
this is not the way to declare dependencies.

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/CharacterDao.kt:
```diff
@@ -0,0 +1,15 @@
+@Dao
+interface CharacterDao {
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
+    suspend fun insertAll(vararg characters: Character)
```
we talked specifically about this vararg, change it for a list

YagoAccenture replied:
Fixed in next commit

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/CharacterDto.kt:
```diff
@@ -0,0 +1,24 @@
+
+data class UbicationDto(
+    val name: String,
+    val url: String
```
you are not using this url.
This makes you create a nested object, now you have mappers, you can remove the ubication object after parse it from remote and work with a simple originName: String and locationName: String.

YagoAccenture replied:
Fixed in next commit, the url variable has been removed

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/DetailsViewModel.kt:
```diff
@@ -1,30 +1,32 @@
+    private suspend fun fetchFirstEpisodeDetails(episodeUrl: String) {
+        try {
+            val episodeResult = repository.fetchEpisodeDetails(episodeUrl)
+            episodeResult.fold(
```
what does the .fold do?

YagoAccenture replied:
It is a function, to handle a result, depending on whether it is successful or failed, it will be modified to use if

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/MainActivity.kt:
```diff
@@ -12,10 +12,7 @@ import com.google.android.material.snackbar.Snackbar
-    private val repository = RickAndMortyRepository(apiRick)
-    private val viewModel: MainViewModel = MainViewModel(this, repository)
-
+    private lateinit var viewModel: MainViewModel
```
what we talked about lateinits?

YagoAccenture replied:
Fixed in next commit.

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/RickAndMortyRepository.kt:
```diff
@@ -1,43 +1,55 @@
+import java.io.IOException
 
-class RickAndMortyRepository(private val apiRick: ApiRick) {
+class RickAndMortyRepository(private val apiRick: ApiRick, private val characterDao: CharacterDao) {
```
we talked about this. Here you need to use the source of the data, from remote you are using the Api, from local you are using the DB, you shouldn't use the Dao. Keep the consistency.

YagoAccenture replied:
Fixed in next commit.

Commit:
Step 11:
Removed lateinit

Commit:
Step 11 EDIT:
vararg, changed for a list

Commit:
Step 11 EDIT:
Ensure data persistence and fallback with local database via CharacterDao.

Commit:
Step 11 EDIT:
Replaced fold with if statements for clearer success/failure handling

Commit:
Step 11 EDIT:
Dependency implementation fix.
Now the primarykey is the character id.
Removed duplicate viewModel.

Commit:
Step 11 EDIT:
Fix character mapping and storage logic to ensure proper data loading from API.

Commit:
Step 11 DELETE:
The url variable has been removed.

Commit:
Step 7 EDIT: Implement data sources for remote API and local DB access consistency

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/RickAndMortyRepository.kt:
```diff
@@ -1,30 +1,61 @@
+                        val locationName = dto.location?.name ?: ""
+                        val firstEpisode = dto.episode.firstOrNull() ?: ""
+
+                        Character(
```
move this to a mapper

YagoAccenture replied:
Solved in next commit.

Commit:
Step 11 EDIT:
Created mappers for episode and character.
Removed threads within threads.

Commit:
Step 11 EDIT:
Created entity for episode, with its mapper, and code refactoring

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/CharacterRemoteDataSource.kt:
```diff
@@ -0,0 +1,8 @@
+
+import retrofit2.Response
+
+interface CharacterRemoteDataSource {
```
One of the methods is returning a Response and the other one the List, seems like Coderabbit is right here.

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/DetailsActivity.kt:
```diff
@@ -18,14 +20,15 @@ class DetailsActivity : AppCompatActivity(), DetailsViewTranslator {
         setContentView(R.layout.details_screen)
 
+        val db = AppDatabase.getDatabase(this)
         val apiRick: ApiRick = ApiService.retrofit.create(ApiRick::class.java)
-        val repository = RickAndMortyRepository(apiRick)
+        val remoteDataSource = RemoteDataSourceImpl(apiRick)
+        val localDataSource = LocalDataSourceImpl(AppDatabase.getDatabase(this).characterDao())
+        val repository = RickAndMortyRepository(remoteDataSource, localDataSource)
```
true, @YagoAccenture

YagoAccenture replied:
Solved in next commit.

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/MainActivity.kt:
```diff
@@ -27,7 +24,15 @@ class MainActivity : AppCompatActivity(), ViewTranslator {
         setContentView(R.layout.activity_main)
 
-        viewModel.onCreate()
+        val database = AppDatabase.getDatabase(this)
+        val characterDao = database.characterDao()
+        val apiRick: ApiRick = ApiService.retrofit.create(ApiRick::class.java)
+        val remoteDataSource = RemoteDataSourceImpl(apiRick)
+        val localDataSource = LocalDataSourceImpl(AppDatabase.getDatabase(this).characterDao())
+        val repository = RickAndMortyRepository(remoteDataSource, localDataSource)
+
```
same here

YagoAccenture replied:
Solved in next commit.

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/RemoteDataSourceImpl.kt:
```diff
@@ -0,0 +1,32 @@
+                return responseBody.results.mapToEntityList()
+            }
+        }
+        throw Exception("")
```
Send a proper message here

YagoAccenture replied:
Solved in next commit

PabloGrelaM commented on app/src/main/java/com/mstudio/superheroarch/RemoteDataSourceImpl.kt:
```diff
@@ -0,0 +1,32 @@
+                return Response.success(episodeEntity)
+            }
+        }
+        throw Exception("")
```
Rick is right

YagoAccenture replied:
Solved in next commit

Commit:
Step 11 EDIT:
Bug corrections
